pythnet/remote-executor/programs/remote-executor/src/lib.rs

```diff
@@ -1,15 +1,61 @@
-use anchor_lang::prelude::*;
+#![deny(warnings)]
+
+use anchor_lang::{prelude::*, solana_program::borsh::get_packed_len};
+use state::{claim_record::ClaimRecord, posted_vaa::PostedVaa};
+
+mod state;
+mod error;
 
 declare_id!("Fg6PaFpoGXkYsidMpWTK6W2BeZ7FEfcYkg476zPFsLnS");
 
 #[program]
 pub mod remote_executor {
+    use anchor_lang::solana_program::{program::invoke_signed, instruction::Instruction};
+
+    use crate::{state::{posted_vaa::CHAIN_ID_SOLANA, governance_payload::ExecutorPayload}, error::ExecutorError};
+
     use super::*;
 
     pub fn execute_posted_vaa(ctx: Context<ExecutePostedVaa>) -> Result<()> {
+        let posted_vaa = &ctx.accounts.posted_vaa;
+        let claim_record = &mut ctx.accounts.claim_record;
+        assert_or_err(posted_vaa.emitter_chain == CHAIN_ID_SOLANA, err!(ExecutorError::EmitterChainNotSolana))?;
+        assert_or_err(posted_vaa.sequence > claim_record.sequence, err!(ExecutorError::NonIncreasingSequence))?;
+        claim_record.sequence = posted_vaa.sequence;
+
+        let payload = ExecutorPayload::try_from_slice(&posted_vaa.payload)?;
+        payload.check_header()?;
+        
+        for instruction in payload.instructions.iter().map(Instruction::from) {
+            // TO DO: We currently pass `remaining_accounts` down to the CPIs, is there a more efficient way to do it?
+            // TO DO: We don't pass ctx.accounts so executor_key needs to be both in the anchor context and in remaining accounts
```

Comment:
just so i'm clear on how this works: the `Instruction` already has a list of accounts that it expects to operate on, and this CPI works as long as `remaining_accounts` is a superset of the accounts in the instruction?

Reply:
also i think passing in the executor_key again as part of the remaining_accounts is more explicit anyway. Adding the executor_key by default here could be confusing to the user.

Reply:
Yes that's exact

Reply:
If you're cool with having the user pass executer_key as part of the additional accounts then this should work.

Reply:
Isn't executor_key passed to the instruction and parsed by anchor (to check seeds)? Then I'm not sure it's included in remaining accounts (there is a account deduplication logic in solana i think)

Reply:
I think there's account deduplication at this level. We'll check in the tests.